File: frontend/src/lib/domain/project/project.ts
/**
 * Project domain types
 * Mirrors backend: internal/domain/project/project.go
 */

import type { Pagination } from '../utils/index.ts';
import type { User } from '../user/index.ts';
import type { ObjectData } from '../facility/index.ts';
import type { Phase } from '../phase/index.ts';
import type { components } from '$lib/api/generated/schema.js';
export type ProjectStatus = 'planned' | 'ongoing' | 'completed';

export interface Project {
  id: string;
  name: string;
  description: string;
  status: ProjectStatus;
  start_date?: string;
  phase_id: string;
  phase?: Phase | null;
  creator_id: string;
  created_at: string;
  updated_at: string;
}

export interface CreateProjectRequest {
  name: string;
  description?: string;
  status?: ProjectStatus;
  start_date?: string;
  phase_id: string;
}

/** API DTO for PATCH /projects/{id}; the ID belongs exclusively in the path. */
export type UpdateProjectRequest =
  components['schemas']['github_com_besart951_go_infra_link_backend_internal_handler_dto_project.UpdateProjectRequest'];

export interface ProjectListParams {
  page?: number;
  limit?: number;
  search?: string;
  status?: ProjectStatus;
  phase_id?: string;
}

export interface ProjectListResponse extends Pagination<Project> {}

export interface ProjectUserListResponse {
  items: User[];
}

export interface ProjectObjectDataListResponse extends Pagination<ObjectData> {}

File: backend/internal/handler/dto/project.go
package dto

import (
	"time"

	"github.com/besart951/go_infra_link/backend/internal/domain/project"
	"github.com/google/uuid"
)

// Project DTOs

type CreateProjectRequest struct {
	Name        string     `json:"name" binding:"required,min=1,max=255"`
	Description string     `json:"description"`
	Status      string     `json:"status" binding:"omitempty,oneof=planned ongoing completed"`
	StartDate   *time.Time `json:"start_date"`
	PhaseID     uuid.UUID  `json:"phase_id" binding:"required"`
}

type UpdateProjectRequest struct {
	Name        string                `json:"name" binding:"omitempty,min=1,max=255"`
	Description string                `json:"description"`
	Status      project.ProjectStatus `json:"status" binding:"omitempty,oneof=planned ongoing completed"`
	StartDate   *time.Time            `json:"start_date"`
	PhaseID     *uuid.UUID            `json:"phase_id"`
}

type ProjectResponse struct {
	ID          uuid.UUID             `json:"id"`
	Name        string                `json:"name"`
	Description string                `json:"description"`
	Status      project.ProjectStatus `json:"status"`
	StartDate   *time.Time            `json:"start_date"`
	PhaseID     uuid.UUID             `json:"phase_id"`
	CreatorID   uuid.UUID             `json:"creator_id"`
	CreatedAt   time.Time             `json:"created_at"`
	UpdatedAt   time.Time             `json:"updated_at"`
}

type ProjectListResponse struct {
	Items      []ProjectResponse `json:"items"`
	Total      int64             `json:"total"`
	Page       int               `json:"page"`
	TotalPages int               `json:"total_pages"`
}
